drivers/rtc/rtc_ll_stm32.c

```diff
@@ -322,6 +576,482 @@ static int rtc_stm32_get_time(const struct device *dev, struct rtc_time *timeptr
+		p_rtc_alrm = &(data->rtc_alrm_b);
+	} else {
+		LOG_ERR("invalid alarm ID %d", id);
+		err = -EINVAL;
+		goto unlock;
+	}
+
+	p_ll_rtc_alarm = &(p_rtc_alrm->ll_rtc_alrm);
+	p_ll_rtc_alrm_time = &(p_ll_rtc_alarm->AlarmTime);
+
+	memset(timeptr, -1, sizeof(struct rtc_time));
+
+#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP) || defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
+	LL_PWR_EnableBkUpAccess();
+#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPCR_DBP || PWR_DBPR_DBP */
+
+	if (id == RTC_STM32_ALRM_A) {
+		timeptr->tm_sec = bcd2bin(LL_RTC_ALMA_GetSecond(RTC));
+		timeptr->tm_min = bcd2bin(LL_RTC_ALMA_GetMinute(RTC));
+		timeptr->tm_hour = bcd2bin(LL_RTC_ALMA_GetHour(RTC));
+		timeptr->tm_wday = bcd2bin(LL_RTC_ALMA_GetWeekDay(RTC));
+		timeptr->tm_mday = bcd2bin(LL_RTC_ALMA_GetDay(RTC));
+		ll_mask = LL_RTC_ALMA_GetMask(RTC);
+	} else {
+		timeptr->tm_sec = bcd2bin(LL_RTC_ALMB_GetSecond(RTC));
+		timeptr->tm_min = bcd2bin(LL_RTC_ALMB_GetMinute(RTC));
+		timeptr->tm_hour = bcd2bin(LL_RTC_ALMB_GetHour(RTC));
+		timeptr->tm_wday = bcd2bin(LL_RTC_ALMB_GetWeekDay(RTC));
+		timeptr->tm_mday = bcd2bin(LL_RTC_ALMB_GetDay(RTC));
+		ll_mask = LL_RTC_ALMB_GetMask(RTC);
+	}
+
+#if defined(PWR_CR_DBP) || defined(PWR_CR1_DBP) || defined(PWR_DBPCR_DBP) || defined(PWR_DBPR_DBP)
+	LL_PWR_DisableBkUpAccess();
+#endif /* PWR_CR_DBP || PWR_CR1_DBP || PWR_DBPCR_DBP || PWR_DBPR_DBP */
+
+	p_rtc_alrm->user_mask = 0;
+	if (id == RTC_STM32_ALRM_A) {
+		if ((ll_mask & LL_RTC_ALMA_MASK_SECONDS) == 0x0) {
+			p_rtc_alrm->user_mask = RTC_ALARM_TIME_MASK_SECOND;
+		}
+		if ((ll_mask & LL_RTC_ALMA_MASK_MINUTES) == 0x0) {
+			p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_MINUTE;
+		}
+		if ((ll_mask & LL_RTC_ALMA_MASK_HOURS) == 0x0) {
+			p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_HOUR;
+		}
+		if ((ll_mask & LL_RTC_ALMA_MASK_DATEWEEKDAY) == 0x0) {
+			if (timeptr->tm_wday) {
+				p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_WEEKDAY;
+			} else {
+				p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_MONTHDAY;
+			}
+		}
+	} else {
+		if ((ll_mask & LL_RTC_ALMB_MASK_SECONDS) == 0x0) {
+			p_rtc_alrm->user_mask = RTC_ALARM_TIME_MASK_SECOND;
+		}
+		if ((ll_mask & LL_RTC_ALMB_MASK_MINUTES) == 0x0) {
+			p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_MINUTE;
+		}
+		if ((ll_mask & LL_RTC_ALMB_MASK_HOURS) == 0x0) {
+			p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_HOUR;
+		}
+		if ((ll_mask & LL_RTC_ALMB_MASK_DATEWEEKDAY) == 0x0) {
+			if (timeptr->tm_wday) {
+				p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_WEEKDAY;
+			} else {
+				p_rtc_alrm->user_mask |= RTC_ALARM_TIME_MASK_MONTHDAY;
+			}
+		}
+	}
+	*mask = p_rtc_alrm->user_mask;
+
+	LOG_DBG("get alarm: mday = %d, wday = %d, hour = %d, min = %d, sec = %d, "
+		"mask = 0x%04x", timeptr->tm_mday, timeptr->tm_wday, timeptr->tm_hour,
+		timeptr->tm_min, timeptr->tm_sec, *mask);
+
+unlock:
+	k_mutex_unlock(&data->lock);
+
+	return err;
+}
+
+static int rtc_stm32_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
```

Comment:
This function is very long, is there a way to split it in smaller pieces?

Reply:
Fixed